### PR TITLE
hw11

### PR DESCRIPTION
**Pull request by ZubenkoHalyna**

### CODE REVIEW

File: src/main/java/mate/academy/internetshop/dao/jdbc/JdbcOrderDao.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        }
+    }
+
+    private Order getOrder(Long id, Long userId) throws SQLException {
+        return new Order(id, productDao.getByOrder(id),
+                userDao.get(userId).orElseThrow());
+    }
+
+    private boolean saveProducts(Order order, Connection con) throws SQLException {
+        String query = "INSERT INTO orders_products VALUES (?, ?)";
```
Comment (dmytro-hryhoruk) [inline src/main/java/mate/academy/internetshop/dao/jdbc/JdbcOrderDao.java]: it's better to use insert with field's names in case you will change the table(order of field etc.) 
"INSERT INTO orders_products(order_id, product_id) VALUES (?, ?)"

File: src/main/java/mate/academy/internetshop/dao/jdbc/JdbcOrderDao.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+    @Override
+    public List<Order> getByUser(Long userId) {
+        try (Connection con = ConnectionUtil.getConnection()) {
+            String query = "SELECT order_id FROM orders WHERE user_id = ?";
+            PreparedStatement statement = con.prepareStatement(query);
+            statement.setLong(1, userId);
+            ResultSet resultSet = statement.executeQuery();
+            List<Order> orders = new ArrayList<>();
+            User user = userDao.get(userId).orElseThrow();
```
Comment (dmytro-hryhoruk) [inline src/main/java/mate/academy/internetshop/dao/jdbc/JdbcOrderDao.java]: What for do we need this user?

File: src/main/java/mate/academy/internetshop/dao/jdbc/JdbcProductDao.java [hunk 3]
```diff
@@ -49,6 +49,7 @@ public Product create(Product product) {
             }
             return Optional.empty();
         } catch (SQLException e) {
+            LOGGER.error(e);
             throw new RuntimeException(e);
```
Comment (dmytro-hryhoruk) [inline src/main/java/mate/academy/internetshop/dao/jdbc/JdbcProductDao.java]: we don't have log exceptions separately cause they will be logged anyway

File: src/main/java/mate/academy/internetshop/dao/jdbc/JdbcOrderDao.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+            return statement.executeUpdate() > 0;
+        } catch (SQLException e) {
+            LOGGER.error(e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    private Order getOrder(Long id, Long userId) throws SQLException {
+        return new Order(id, productDao.getByOrder(id),
+                userDao.get(userId).orElseThrow());
```
Comment (dmytro-hryhoruk) [inline src/main/java/mate/academy/internetshop/dao/jdbc/JdbcOrderDao.java]: you have an option to replace User user in ShoppingCart and Order, to Long userId. This way you will be able to remove usage of userDao from others Daos

File: src/main/java/mate/academy/internetshop/dao/jdbc/JdbcProductDao.java [hunk 5]
```diff
@@ -79,14 +73,8 @@ public Product update(Product product) {
             statement.setString(1, product.getName());
             statement.setBigDecimal(2, product.getPrice());
             statement.setLong(3, product.getId());
-            if (statement.executeUpdate() > 0) {
-                return get(product.getId()).get();
-            }
-            String msg = "Enable to update product with id " + product.getId();
-            LOGGER.error(msg);
-            throw new RuntimeException(msg);
```
Comment (dmytro-hryhoruk) [inline src/main/java/mate/academy/internetshop/dao/jdbc/JdbcProductDao.java:85]: why did you remove executeUpdate?

File: src/main/java/mate/academy/internetshop/dao/jdbc/JdbcOrderDao.java (hunk 4 above)
Comment (ZubenkoHalyna) [reply src/main/java/mate/academy/internetshop/dao/jdbc/JdbcOrderDao.java:131]: In this case I should find user by Id in GetOrderController, GetUserOrdersController, GetShoppingCartController and CompleteOrderController. If I may I choose to leave it as it is.

File: src/main/java/mate/academy/internetshop/dao/jdbc/JdbcOrderDao.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+            return orders;
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public Order update(Order order) {
+        try (Connection con = ConnectionUtil.getConnection()) {
+            String query = "UPDATE orders SET user_id = ? WHERE order_id = ?";
```
Comment (dmytro-hryhoruk) [inline src/main/java/mate/academy/internetshop/dao/jdbc/JdbcOrderDao.java]: In our case update order means update it's products. I'm not sure that we will pass order from one user to another in our internetshop

File: src/main/java/mate/academy/internetshop/dao/jdbc/JdbcOrderDao.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+            PreparedStatement statement = con.prepareStatement(query);
+            statement.setLong(1, order.getUser().getId());
+            statement.setLong(2, order.getId());
+            statement.executeUpdate();
+            query = "DELETE FROM orders_products WHERE order_id = ?";
+            statement = con.prepareStatement(query);
+            statement.setLong(1, order.getId());
+            statement.executeUpdate();
+            saveProducts(order, con);
+            return get(order.getId()).get();
```
Comment (dmytro-hryhoruk) [inline src/main/java/mate/academy/internetshop/dao/jdbc/JdbcOrderDao.java]: do you need to use get here?

File: src/main/java/mate/academy/internetshop/dao/jdbc/JdbcOrderDao.java [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    @Override
+    public Order update(Order order) {
+        try (Connection con = ConnectionUtil.getConnection()) {
+            String query = "UPDATE orders SET user_id = ? WHERE order_id = ?";
+            PreparedStatement statement = con.prepareStatement(query);
+            statement.setLong(1, order.getUser().getId());
+            statement.setLong(2, order.getId());
+            statement.executeUpdate();
+            query = "DELETE FROM orders_products WHERE order_id = ?";
+            statement = con.prepareStatement(query);
```
Comment (dmytro-hryhoruk) [inline src/main/java/mate/academy/internetshop/dao/jdbc/JdbcOrderDao.java]: we should close our statement too

File: src/main/java/mate/academy/internetshop/dao/jdbc/JdbcProductDao.java [hunk 9]
```diff
@@ -72,17 +68,12 @@ public Product create(Product product) {
...
+            return get(product.getId()).get();
```
Comment (dmytro-hryhoruk) [inline src/main/java/mate/academy/internetshop/dao/jdbc/JdbcProductDao.java]: you already have product  that passed into your method with a new values inside

File: src/main/java/mate/academy/internetshop/dao/jdbc/JdbcUserDao.java [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+            PreparedStatement statement = conn.prepareStatement(query);
+            statement.setString(1, user.getName());
+            statement.setString(2, user.getLogin());
+            statement.setString(3, user.getPassword());
+            statement.setLong(4, user.getId());
+            statement.executeUpdate();
+            query = "DELETE FROM users_roles WHERE user_id = ?";
+            statement = conn.prepareStatement(query);
+            statement.setLong(1, user.getId());
+            saveRoles(user, conn);
```
Comment (dmytro-hryhoruk) [inline src/main/java/mate/academy/internetshop/dao/jdbc/JdbcUserDao.java:107]: What if we decided to remove some roles from user and then update him?

File: src/main/java/mate/academy/internetshop/dao/jdbc/JdbcShoppingCartDao.java [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public ShoppingCart update(ShoppingCart shoppingCart) {
+        try (Connection con = ConnectionUtil.getConnection()) {
+            String query = "UPDATE shopping_carts SET user_id = ? "
+                    + "WHERE shopping_cart_id = ?";
```
Comment (dmytro-hryhoruk) [inline src/main/java/mate/academy/internetshop/dao/jdbc/JdbcShoppingCartDao.java]: same as with order

File: src/main/java/mate/academy/internetshop/dao/jdbc/JdbcOrderDao.java (hunk 4 above)
Comment (dmytro-hryhoruk) [reply src/main/java/mate/academy/internetshop/dao/jdbc/JdbcOrderDao.java]: Yes, and if you think about if you have method where you should find something by user, it's not too hard to modify this method to find something by userId. It's up to you after all)

File: src/main/java/mate/academy/internetshop/dao/jdbc/JdbcProductDao.java (hunk 9 above)
Comment (ZubenkoHalyna) [reply src/main/java/mate/academy/internetshop/dao/jdbc/JdbcProductDao.java]: price field may change because of type convercion after saving to database

Comment (dmytro-hryhoruk) [reply src/main/java/mate/academy/internetshop/dao/jdbc/JdbcProductDao.java]: what type you are using for price column?

Comment (ZubenkoHalyna) [reply src/main/java/mate/academy/internetshop/dao/jdbc/JdbcProductDao.java]: BigDecimal in java and Decimal(10, 0) in SQL

File: src/main/java/mate/academy/internetshop/dao/jdbc/JdbcUserDao.java (hunk 10 above)
Comment (ZubenkoHalyna) [reply src/main/java/mate/academy/internetshop/dao/jdbc/JdbcUserDao.java:107]: A missed executeUpdate, I understand.

File: src/main/java/mate/academy/internetshop/dao/jdbc/JdbcProductDao.java (hunk 9 above)
Comment (ZubenkoHalyna) [reply src/main/java/mate/academy/internetshop/dao/jdbc/JdbcProductDao.java]: I've deleted that code, but I think it has some sense, but we in any case dosen't use value that returned by update method

Comment (dmytro-hryhoruk) [reply src/main/java/mate/academy/internetshop/dao/jdbc/JdbcProductDao.java]: try to use Decimal(10,2) , that should reduce the difference

**Review by dmytro-hryhoruk — APPROVED** [review]
